### PR TITLE
Add logs button on the trace page.

### PR DESCRIPTION
**Pull request by fedj**
When troubleshooting our system, we often need to check
if there were logs. Since a good pratice is to add the
trace id to your logs, it seems really natural to have
a button to open a link to query your logs with this id.
Also, people generally have kibana to search for specific
logs (or at least have a query-able log service).

The button is deactivated by default (if you don't have a log
query service or the trace id in the logs) and it's simply
activated by adding LOGS_URL="http://company.com/logs?query={traceId}".
The traceId will be contextually replaced.

Here's a preview:
logs button

### CODE REVIEW

**Comment by codefromthecrypt** [comment]
cool start! wanna add a screen shot?

cc @openzipkin/elasticsearch @rogeralsing @eirslett @basvanbeek

**Comment by codefromthecrypt** [comment]
cc'ing other folks who have interest in zipkin + kibana @marcingrzejszczak @dragontree101 @fedj @dan-tr @C-Otto

**Comment by C-Otto** [comment]
This seems to solve a similar issue as described in #1300

File: zipkin-server/src/main/resources/zipkin-server.yml [hunk 1]
```diff
@@ -111,7 +111,9 @@ zipkin:
     # - .*example2.com
     # Default is "match all websites"
     instrumented: .*
-
+    # Url of the logs associated with a given request.
+    # It will be formatted with the trace id if your url contains ${traceId}} (e.g. http://kibana/trace${traceId})
```
Comment (C-Otto) [inline zipkin-server/src/main/resources/zipkin-server.yml]: There are two } instead of one

File: zipkin-ui/js/component_data/trace.js [hunk 2]
```diff
@@ -5,11 +5,14 @@ import traceToMustache from '../../js/component_ui/traceToMustache';
 
 export default component(function TraceData() {
   this.after('initialize', function() {
-    $.ajax(`/api/v1/trace/${this.attr.traceId}`, {
+    const traceId = this.attr.traceId;
+    const logsUrlNonFormatted = this.attr.logsUrl;
+    const logsUrl = logsUrlNonFormatted.replace('{traceId}', traceId);
```
Comment (C-Otto) [inline zipkin-ui/js/component_data/trace.js]: I don't know JavaScript enough to be sure, but I assume that `${traceId}` and `{traceId}` are different when replacing?

**Comment by C-Otto** [comment]
I like the idea! A screenshot would help a bit.

File: zipkin-server/src/main/resources/zipkin-server.yml [hunk 3]
```diff
@@ -111,7 +111,9 @@ zipkin:
...
+    logsUrl: ${LOGS_URL:}
```
Comment (shakuzen) [inline zipkin-server/src/main/resources/zipkin-server.yml]: Small nit-pick: other variables are kebab-case in this file, so it might match the style better if this were too. Thanks to Spring Boot's relaxed property binding, it shouldn't effectively matter either way, though.

**Review by shakuzen — CHANGES_REQUESTED** [review]
This sounds like a nice enhancement. Thanks for contributing. Adding a couple of tests for this new behavior would be extra nice.

1) Verify the absence of the button when the property is not set
2) Verify the presence of the button with the correctly replaced URL when the property is set.

**Comment by fedj** [comment]
@C-Otto done
@shakuzen I'm not really familiar with javascript (actually not at all), would you have a pointer on how to do these kind of unit tests ?

File: zipkin-autoconfigure/ui/src/main/java/zipkin/autoconfigure/ui/ZipkinUiProperties.java [hunk 4]
```diff
@@ -22,6 +22,7 @@
   private int queryLimit = 10;
   private int defaultLookback = (int) TimeUnit.DAYS.toMillis(7);
   private String instrumented = ".*";
+  private String logsUrl = "";
```
Comment (codefromthecrypt) [inline zipkin-autoconfigure/ui/src/main/java/zipkin/autoconfigure/ui/ZipkinUiProperties.java]: I'd initialize this to null, so that you don't end up with an empty string in the json

File: zipkin-autoconfigure/ui/src/main/java/zipkin/autoconfigure/ui/ZipkinUiProperties.java [hunk 5]
```diff
@@ -54,4 +55,12 @@ public String getInstrumented() {
   public void setInstrumented(String instrumented) {
     this.instrumented = instrumented;
   }
+
+  public String getLogsUrl() {
+    return logsUrl;
+  }
+
+  public void setLogsUrl(String logsUrl) {
```
Comment (codefromthecrypt) [inline zipkin-autoconfigure/ui/src/main/java/zipkin/autoconfigure/ui/ZipkinUiProperties.java:64]: set if not null or empty

Comment (codefromthecrypt) [reply zipkin-autoconfigure/ui/src/main/java/zipkin/autoconfigure/ui/ZipkinUiProperties.java:64]: I'd test this in ZipkinUiAutoConfigurationTest basically that when `zipkin.ui.logs-url:` (empty string), `getLogsUrl()` is null, and when it is set to a url, that's readable as `getLogsUrl()`. This helps make sure there's no typos hard to see.

File: zipkin-ui/js/component_data/trace.js [hunk 6]
```diff
@@ -3,13 +3,22 @@ import $ from 'jquery';
 import {getError} from '../../js/component_ui/error';
 import traceToMustache from '../../js/component_ui/traceToMustache';
 
+function toContextualLogsUrl(logsUrl, traceId) {
+  if (logsUrl !== undefined) {
```
Comment (codefromthecrypt) [inline zipkin-ui/js/component_data/trace.js]: I think you can shorten this to `if (logsUrl)` ?

Comment (codefromthecrypt) [reply zipkin-ui/js/component_data/trace.js]: if you export this function, you should be able to test it by making a class test/component_data/trace.test.js 

something like... (please use the exact format of kibana search url you use as that helps folks understand how it is used)
```javascript
import {toContextualLogsUrl} from '../../js/component_data/trace';

describe('toContextualLogsUrl', () => {
  it('replaces token in logsUrl when set', () => {
    toContextualLogsUrl('http://some-realistic-logs-url/{traceId}', 'abcd').should.equal(''http://some-realistic-logs-url/abcd'');
  });
});
```

File: zipkin-ui/test/component_ui/traceToMustache.test.js [hunk 7]
```diff
@@ -40,20 +40,21 @@ const span3 = span(12345, 'methodcall2', span3Id, span2Id, 300, 50, annotations3
 const span4 = span(12345, 'methodcall2', span4Id, span3Id, 400, 100, annotations4);
 
 const trace = [span1, span2, span3, span4];
+const logsUrl = '';
```
Comment (codefromthecrypt) [inline zipkin-ui/test/component_ui/traceToMustache.test.js]: I would make this method scope, and set it to undefined as I think that's what it would be in practice.
then, add a test for what happens when it isn't undefined. For example, traceToMustache result includes the logsUrl

File: zipkin-ui/templates/trace.mustache [hunk 8]
```diff
@@ -4,6 +4,9 @@
     <li class=''><a href='#'><strong>Services:</strong> <span class='badge'>{{services}}</span></a></li>
     <li class=''><a href='#'><strong>Depth:</strong> <span class='badge'>{{depth}}</span></a></li>
     <li class=''><a href='#'><strong>Total Spans:</strong> <span class='badge'>{{totalSpans}}</span></a></li>
+    {{#logsUrl}}
```
Comment (codefromthecrypt) [inline zipkin-ui/templates/trace.mustache:7]: not sure there's an existing class to cover this, so one would need to be made.. I can help think of one, or maybe the notes I've made so far can get you there?

**Review by codefromthecrypt — COMMENTED** [review]
I'm fairly novice at javascript, too, but hopefully the notes help. There's instructions on how to run tests in the zipkin-ui README iirc.

It would be nice to see how it looks when someone clicks the link, too. Ex does it pop out, etc? feel free to use something like OSX preview, tools, annotate to black-out anything sensitive.

File: zipkin-ui/templates/trace.mustache (hunk 8 above)
Comment (fedj) [reply zipkin-ui/templates/trace.mustache:7]: Done for all but for the mustache test. I'm not sure to see how to do it properly here.

**Comment by codefromthecrypt** [comment]
@haf what do you think about something like this and logary?

**Comment by codefromthecrypt** [comment]
@fedj mind sharing a screen shot of the logs detail? is it a pop-out or a new tab?

**Comment by haf** [comment]
@adriancole I think it would be excellent. We have a Kibana dashboard at qvitoo and frequently need to correlate logs. Logary does all of the above collecting too. I'm working through the https://github.com/logary/logary/issues and have been for a while now; the next two parts is creating a complete C# facade like the existing Logary.Facade, improving how derived metrics are configured and then create examples including ones with Zipkin in them. This would enable e.g. sending a snapshot 1 minute rolling average together with the span, together with the annotations, that are sent to Zipkin.

However, Logary's API is now well geared towards the logging structure needed for Zipkin; hierarchical passing-down of loggers, applying middleware on the fly (with TimeScope), or alternatively correlating span annotations in the Logary Zipkin target. So the extra work needed to fully support Zipkin now, is very small.

**Comment by fedj** [comment]
@adriancole Currently it's just a link so it depends on your browser settings (I guess). I think it will open the link in the same tab by default (at least this is what it does on Safari without previewing it)

**Comment by codefromthecrypt** [comment]
thanks for the help on this. It will give folks a way to experiment with log correlation without visible impact to those who aren't interested.

**Comment by codefromthecrypt** [comment]
if you don't mind, please verify the snapshot exec jar acts as expected!
http://oss.jfrog.org/oss-snapshot-local/io/zipkin/java/zipkin-server/1.16.3-SNAPSHOT/

oh and probably we need to update zipkin-server/README.md with details on this feature.

**Comment by jcarres-mdsol** [comment]
I think this PR lacked documentation on how to use this new feature.

**Comment by codefromthecrypt** [comment]
yeah noticed that post-merge, too. we need to add the description of this PR or similar to zipkin-server/README.md

**Comment by codefromthecrypt** [comment]
actually, the detailed stuff from the description should go to zipkin-ui/README.md and only description of the env variable to zipkin-server/README.md that way the latter doesn't grow endlessly

**Comment by fedj** [comment]
@jcarres-mdsol, @adriancole You're both right, I'll do that tomorrow if you don't mind

**Comment by fedj** [comment]
@adriancole It's working as expected for the last jar (-6-exec.jar).
I didn't forget about doc, will do another PR today.

**Comment by codefromthecrypt** [comment]
:thumbsup: thanks for the follow-up

**Comment by codefromthecrypt** [comment]
shortly, brave will include SLF4J support which could integrate with this https://github.com/openzipkin/brave/pull/389
